### ISSUE TITLE
DeviceClient keep trying to connect to IoTHub causing lots of Unauthorized errors

### ISSUE THREAD

**Issue by kristofferpagels** [Azure/azure-iot-sdk-csharp#886]
-**OS, version, SKU and CPU architecture used:** Windows 10 Desktop x64
-**Application's .NET Target Framework :** net472
-**Device:** Laptop
-**SDK version used:** 1.19.0
-**TransportType:** AMQP

## Description of the issue:
If the amqp connection is not closed correctly, and the device keys are changed, then the connection will be kept open, and at some point in time it will loop and try to reconnect with the IoTHub causing lots of Unauthorized errors (in the range 40-600 per second). If this happens, the client is not notified, and no exceptions are thrown on the client side, so there is no way to know that the system is in an invalid state.

## Steps to reproduce
0. In the Azure Portal -> IoTHub -> Monitoring -> Diagnostic settings, enable Diagnostics Settings for Connections and log to Log Analytics.
0. Run the program 
0. Wait... In my tests, around 20-30 minutes should do it.
0. In the Azure Portal -> IoTHub -> Monitoring -> Logs, enter the following query. 

Query:
```
AzureDiagnostics 
| where TimeGenerated > ago(2h)
| where Category == "Connections"
| where ResultDescription == "Unauthorized"
| summarize events_count=count(Category) by bin(TimeGenerated, 1m) 
| order by TimeGenerated asc 
```
Result:
```
2019-04-01T08:26:00.000 | 3
2019-04-01T08:27:00.000 | 4
2019-04-01T09:17:00.000 | 1,394  
2019-04-01T09:18:00.000 | 2,625
2019-04-01T09:19:00.000 | 2,613
```
As this is a random failure, you might have to retry and run the program multiple times.
As it can be seen, the number of unauthorized errors are very high, and no errors are propagated to the client.

Example log from Azure Diagnostics:
```
{
	"TenantId": "****",
	"SourceSystem": "Azure",
	"TimeGenerated [UTC]": "2019-04-01T06:27:07Z",
	"ResultDescription": "Unauthorized",
	"ResourceId": "****",
	"OperationName": "deviceDisconnect",
	"Category": "Connections",
	"Level": "Error",
	"properties_s": "{"deviceId":"****","protocol":"Amqp","authType":null,"maskedIpAddress":"****.XXX","statusCode":"401"}",
	"location_s": "westeurope",
	"SubscriptionId": "****",
	"ResourceGroup": "****",
	"ResourceProvider": "MICROSOFT.DEVICES",
	"Resource": "****",
	"ResourceType": "IOTHUBS",
	"Type": "AzureDiagnostics"
}
```

## Code sample exhibiting the issue:

```
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Azure.Devices;
using Microsoft.Azure.Devices.Client;
using Message = Microsoft.Azure.Devices.Client.Message;
using TransportType = Microsoft.Azure.Devices.Client.TransportType;

namespace ConnectionClosePoC
{
    class Program
    {
        static async Task Main(string[] args)
        {
            string iothubConnectionString = "HostName=myhub.azure-devices.net;SharedAccessKeyName=registryReadWrite;SharedAccessKey=****";

            try
            {
                await Test_Disconnect(iothubConnectionString, "mydevice", "myhub.azure-devices.net");
            }
            catch (Exception e)
            {
                LogLine(e.Message);
            }

            LogLine("DONE");
            Console.ReadKey();
        }

        private static async Task Test_Disconnect(string ioTHubConnectionString, string deviceId, string ioTHubHostName)
        {
            using (var registryManager = RegistryManager.CreateFromConnectionString(ioTHubConnectionString))
            {
                var device = await registryManager.GetDeviceAsync(deviceId);

                if (device == null)
                {
                    device = await registryManager.AddDeviceAsync(new Device(deviceId));
                }

                var key = device.Authentication.SymmetricKey.PrimaryKey;
                var hostname = ioTHubHostName;
                var connectionString = string.Format(CultureInfo.InvariantCulture, "HostName={0};DeviceId={1};SharedAccessKey={2}", hostname, deviceId, key);


                using (var deviceClient = DeviceClient.CreateFromConnectionString(connectionString, TransportType.Amqp))
                {
                    deviceClient.SetConnectionStatusChangesHandler(SomeConnectionStatusChangesHandler);

                    await deviceClient.SetMethodHandlerAsync("someMethod1", SomeMethodCallback, new object());
                    await deviceClient.SetMethodHandlerAsync("someMethod2", SomeMethodCallback, new object());
                    await deviceClient.SetMethodDefaultHandlerAsync(MethodDefaultHandlerCallback, new object());

                    LogLine("Starting recveive loop");
                    var count = 0;
                    var exceptionCount = 0;
                    bool goOn = true;
                    do
                    {
                        await Task.Delay(TimeSpan.FromSeconds(5));

                        Message message = null;
                        try
                        {
                            Log(".");
                            message = await deviceClient.ReceiveAsync(TimeSpan.FromSeconds(20));
                        }
                        catch (Exception e)
                        {
                            exceptionCount++;
                            LogLine(e.Message);

                            if (exceptionCount % 5 == 0)
                            {
                                goOn = false;
                                LogLine("STOP receiving");
                            }
                        }

                        if (count == 1)
                        {
                            LogLine("Change key");
                            device.Status = DeviceStatus.Disabled;
                            device = await registryManager.UpdateDeviceAsync(device);

                            // Wait a bit before enabling again (just to be sure)
                            await Task.Delay(1000);

                            device.Status = DeviceStatus.Enabled;
                            device.Authentication.SymmetricKey.PrimaryKey = GenerateSharedKey(64);
                            device.Authentication.SymmetricKey.SecondaryKey = GenerateSharedKey(64);
                            await registryManager.UpdateDeviceAsync(device);
                        }

                        count++;
                    } while (goOn);

                    LogLine("Ending recveive loop");

                    await deviceClient.SetMethodHandlerAsync("someMethod1", null, null);
                    await deviceClient.SetMethodHandlerAsync("someMethod2", null, null);

                    await deviceClient.CloseAsync();
                    LogLine("Device client closed");

                    LogLine("Starting wait loop");
                    goOn = true;
                    var i = 0;
                    do
                    {
                        await Task.Delay(TimeSpan.FromSeconds(5));
                        i++;

                        if (i == int.MaxValue)
                        {
                            goOn = false;
                            LogLine("STOP: " + i);
                        }
                    } while (goOn);

                    LogLine("Ending wait loop");
                }
            }
        }

        private static void LogLine(string message, params object[] arg)
        {
            Console.WriteLine(DateTime.Now.ToString("s") + " - " + message, arg);
        }

        private static void Log(string message, params object[] arg)
        {
            Console.Write(message, arg);
        }

        private static string GenerateSharedKey(int keyLength)
        {
            var rngCryptoServiceProvider = new RNGCryptoServiceProvider();
            var randomBytes = new byte[keyLength];
            rngCryptoServiceProvider.GetBytes(randomBytes);
            var base64 = Convert.ToBase64String(randomBytes);
            return base64;
        }

        private static Task<MethodResponse> MethodDefaultHandlerCallback(MethodRequest methodRequest, object userContext)
        {
            Console.WriteLine("Received DirectMethodCall for unknown method");
            return Task.FromResult(new MethodResponse(501));
        }

        private static Task<MethodResponse> SomeMethodCallback(MethodRequest methodRequest, object userContext)
        {
            Console.WriteLine("Received DirectMethodCall");
            return Task.FromResult(new MethodResponse(200));
        }

        private static void SomeConnectionStatusChangesHandler(ConnectionStatus status, ConnectionStatusChangeReason reason)
        {
            Console.WriteLine("Status changed to: " + status);
        }
    }
}

```

## Console log of the issue:
Console log:
```
Status changed to: Connected
2019-04-01T08:19:30 - Starting recveive loop
..2019-04-01T08:21:40 - Change key
Status changed to: Disconnected_Retrying
Status changed to: Disconnected
.Status changed to: Connected
....Status changed to: Disconnected_Retrying
Status changed to: Disconnected
.2019-04-01T08:27:03 - Put token failed. status-code: 401, status-description: The specified SAS token has an invalid signature. It does not match either the primary or secondary key..
.2019-04-01T08:27:09 - Put token failed. status-code: 401, status-description: The specified SAS token has an invalid signature. It does not match either the primary or secondary key..
.2019-04-01T08:27:14 - Put token failed. status-code: 401, status-description: The specified SAS token has an invalid signature. It does not match either the primary or secondary key..
.2019-04-01T08:27:20 - Put token failed. status-code: 401, status-description: The specified SAS token has an invalid signature. It does not match either the primary or secondary key..
.2019-04-01T08:27:25 - Put token failed. status-code: 401, status-description: The specified SAS token has an invalid signature. It does not match either the primary or secondary key..
2019-04-01T08:27:25 - STOP receiving
2019-04-01T08:27:25 - Ending recveive loop
2019-04-01T08:27:29 - Device client closed
2019-04-01T08:27:29 - Starting wait loop
```

**Comment by jenoola** [comment]
The issue is fixed in the new release 1.20.0

**Comment by az-iot-builder-01** [comment]
@kristofferpagels, @jenoola, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)
